Sources/NIO/SocketChannel.swift

```diff
@@ -607,18 +606,14 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
 
     internal final func updateCachedAddressesFromSocket(updateLocal: Bool = true, updateRemote: Bool = true) {
         assert(self.eventLoop.inEventLoop)
-        if updateLocal {
-            self.localAddressCached.store(Box(try? self.localAddress0()))
-        }
-        if updateRemote {
-            self.remoteAddressCached.store(Box(try? self.remoteAddress0()))
-        }
+        let local = updateLocal ? try? self.localAddress0() : self.localAddress
+        let remote = updateRemote ? try? self.remoteAddress0() : self.remoteAddress
```

Comment:
There are a bunch of atomic operations hidden in here. This code performs a pair of atomic loads followed by an atomic store, all on the same value. It's very much not clear to me that this is a good performance tradeoff. We probably want to elide the double-load at the very least, because the compiler can't really see past it (these operations are implemented in C and so Swift doesn't really know anything about them).

Reply:
Yeah good point about the double load... Let me fix this.